### PR TITLE
Compute polygon coordinates

### PR DESCRIPTION
**Pull request by comhar**
Changes proposed in this pull request:

- Add a `regular_polygon` method to `ImageDraw`

 ### Usage
##### Draw a red hexagon, rotated by 90 degrees, inscribed by the following bounding box = [(50,50), (150, 150)]
```python
img = Image.new('RGBA', size=(200, 200), color=(255, 0, 0, 0))
draw = ImageDraw.Draw(img)
bbox=[(50,50), (150, 150)]
n_sides=6
draw.regular_polygon(bbox, n_sides, rotation=90, fill='red')
```

~- Add a `compute_polygon_coordinates` method to `ImageDraw`.~

~#### What does `compute_polygon_coordinates` do?~

~`compute_polygon_coordinates` creates a list of coordinates for a polygon of n sides with a given length.~

~### Usage~
~Draw a regular hexagon with side of length `10`, starting at the position `(50, 50)`~

~```python~
~# 1. Compute Co-ordinates~
~coordinates = compute_polygon_coordinates(~
~nb_sides=6,~
~side_length=10,~
~starting_point=(50,50)~
~)~
~# 2. Draw hexagon~
~img = Image.new('RGBA', size=(256, 256), color=(255, 0, 0, 0))~
~draw = ImageDraw.Draw(img)~
~draw.polygon(coordinates, fill=colour)~

### CODE REVIEW

**Comment by radarhere** [comment]
Hmm. What is your use case for this?

Our drawing methods of arc, chord, ellipse, etc, accept a bounding box as an argument, and work within that. So I'm wondering if it would be helpful here to have a `regular_polygon` method, that accepts a bounding box, and the number of sides, and then draws a regular polygon centered within that? Not entirely sure if that's helpful, but I wanted to put the idea out there.

**Comment by nulano** [comment]
It seems to me that either this or @radarhere's suggestion would be a lot more useful if they allowed an arbitrary rotation. It looks like this PR always returns a polygon with a horizontal edge at the bottom.

**Comment by comhar** [comment]
> Hmm. What is your use case for this?

I needed to generate regular polygons for a machine learning project. Although drawing regular polygon's isn't as common as drawing squares or circles, `regular_polygon` would be a nice method to have. 

@radarhere @nulano, I like both of your suggestions. I'll implement in the next iteration.

**Comment by comhar** [comment]
Here's the updated method

```python
# Draw a red hexagon, rotated by 90 degrees, inscribed by the following bounding box = [(50,50), (150, 150)]
img = Image.new('RGBA', size=(200, 200), color=(255, 0, 0, 0))
draw = ImageDraw.Draw(img)
draw.regular_polygon(b_box=[(50,50), (150, 150)], nb_sides=6, rotation=90, fill='red')
```

**Comment by nulano** [comment]
Note that other functions accept both a `[(x0, y0), (x1, y1)]` and `[x0, y0, x1, y1]` bbox in a positional parameter named `xy`, see e.g. https://pillow.readthedocs.io/en/stable/reference/ImageDraw.html#PIL.ImageDraw.ImageDraw.arc

Have you considered what happens if the bbox is not a square? I can't see any tests concerning that. I think the current version draws a polygon bounded by the height of the bbox, possibly spilling over horizontally. I'm wondering if it would be useful to first compute a polygon bounded by the height and then stretch the width to match the bbox (similar to shapes in MS Paint). It would no longer be a regular polygon (not sure how this would affect the function name), but perhaps it would be useful to draw axis-stretched shapes. I'm not sure how common this would be, but in such a rare situation, I imagine it could be quite useful.

File: docs/reference/ImageDraw.rst [hunk 1]
```diff
@@ -254,6 +254,22 @@ Methods
     :param outline: Color to use for the outline.
     :param fill: Color to use for the fill.
 
+
+.. py:method:: ImageDraw.regular_polygon(*, b_box, nb_sides, rotation=0, fill=None, outline=None)
```
Comment (hugovk) [inline docs/reference/ImageDraw.rst:258]: `thing(*, arg1, arg2)` means all the arguments after the asterisk are keyword-only ([PEP 3102](https://www.python.org/dev/peps/pep-3102/)), e.g. the user has to call `thing(arg1=x, arg2=y)`, they cannot call `thing(x, y)`.

We've not required that before, any particular reason to use it here?

Comment (hugovk) [inline docs/reference/ImageDraw.rst:258]: Nits:

Let's call it `bbox` throughout, that's what the rest of the codebase uses.

Also, `n_things` is used quite a lot, and there's very little `nb_things`:

 ```suggestion
.. py:method:: ImageDraw.regular_polygon(*, bbox, n_sides, rotation=0, fill=None, outline=None)
```

Comment (nulano) [reply docs/reference/ImageDraw.rst:258]: I would go with `xy` here, as while most of the code base uses `bbox`, the ImageDraw functions use `xy`: https://pillow.readthedocs.io/en/stable/reference/ImageDraw.html#PIL.ImageDraw.ImageDraw.rectangle

```suggestion
.. py:method:: ImageDraw.regular_polygon(*, xy, n_sides, rotation=0, fill=None, outline=None)
```

Comment (comhar) [reply docs/reference/ImageDraw.rst]: @hugovk , thanks for the review. keyword-only args are a personal preference but I've implemented your suggestion in order to maintain consistency with the rest of the codebase.

**Comment by comhar** [comment]
@nulano , thanks for your review.

> Note that other functions accept both a `[(x0, y0), (x1, y1)]` and `[x0, y0, x1, y1]` bbox in a positional parameter named `xy`, see e.g. https://pillow.readthedocs.io/en/stable/reference/ImageDraw.html#PIL.ImageDraw.ImageDraw.arc

Both formats are now accepted.

> Have you considered what happens if the bbox is not a square? I can't see any tests concerning that. I think the current version draws a polygon bounded by the height of the bbox, possibly spilling over horizontally. I'm wondering if it would be useful to first compute a polygon bounded by the height and then stretch the width to match the bbox (similar to shapes in MS Paint). It would no longer be a regular polygon (not sure how this would affect the function name), but perhaps it would be useful to draw axis-stretched shapes. I'm not sure how common this would be, but in such a rare situation, I imagine it could be quite useful.

I've added a check to ensure the bounding box is square. 
axis-stretched shapes would be a nice addition. As you mentioned, axis stretching would mean the polygon is no longer regular. Perhaps axis-stretching would be better in another PR.

**Comment by comhar** [comment]
@radarhere , @nulano let me know if there's anything I can do to help with the review.

**Comment by hugovk** [comment]
Minor thing, maybe make the filenames a bit shorter? For example:

```diff
-Tests/images/imagedraw_regular_polygon__octagon_with_0_degree_rotation.png
+Tests/images/imagedraw_regular_octagon.png

-Tests/images/imagedraw_regular_polygon__square_with_0_degree_rotation.png
+Tests/images/imagedraw_square.png

-Tests/images/imagedraw_regular_polygon__square_with_45_degree_rotation.png
+Tests/images/imagedraw_square_rotate_45.png
```

File: Tests/test_imagedraw.py [hunk 2]
```diff
@@ -1092,3 +1092,110 @@ def test_same_color_outline():
...
+                (75.0, 6.7),
+                (25.0, 6.7),
+                (0.0, 50.0),
+            ],
+        ),
+    ],
+)
+def test_compute_regular_polygon_vertices(n_polygon_sides, expected_vertices):
+    vertices = ImageDraw._compute_regular_polygon_vertices(
+        n_sides=n_polygon_sides, bbox=[(0, 0), (100, 100)], rotation=0
```
Comment (hugovk) [inline Tests/test_imagedraw.py:1144]:
```suggestion
        n_sides, bbox=[(0, 0), (100, 100)], rotation=0
```

File: Tests/test_imagedraw.py [hunk 3]
```diff
@@ -1092,3 +1092,110 @@ def test_same_color_outline():
...
+                (75.0, 93.3),
+                (100.0, 50.0),
+                (75.0, 6.7),
+                (25.0, 6.7),
+                (0.0, 50.0),
+            ],
+        ),
+    ],
+)
+def test_compute_regular_polygon_vertices(n_polygon_sides, expected_vertices):
```
Comment (hugovk) [inline Tests/test_imagedraw.py:1142]:
```suggestion
def test_compute_regular_polygon_vertices(n_sides, expected_vertices):
```

File: Tests/test_imagedraw.py [hunk 4]
```diff
@@ -1092,3 +1092,110 @@ def test_same_color_outline():
...
+    )
+    draw = ImageDraw.Draw(im)
+    draw.regular_polygon(
+        [(0, 0), (W, H)], n_polygon_sides, rotation=rotation, fill="red"
+    )
+    assert_image_equal(im, Image.open(filename))
+
+
+@pytest.mark.parametrize(
+    "n_polygon_sides, expected_vertices",
```
Comment (hugovk) [inline Tests/test_imagedraw.py:1115]:
```suggestion
    "n_sides, expected_vertices",
```

File: Tests/test_imagedraw.py [hunk 5]
```diff
@@ -1092,3 +1092,110 @@ def test_same_color_outline():
                     operation, mode
                 )
                 assert_image_similar_tofile(im, expected, 1)
+
+
+@pytest.mark.parametrize(
+    "n_polygon_sides, rotation, polygon_name",
```
Comment (hugovk) [inline Tests/test_imagedraw.py:1098]:
```suggestion
    "n_sides, rotation, polygon_name",
```

File: Tests/test_imagedraw.py [hunk 6]
```diff
@@ -1092,3 +1092,110 @@ def test_same_color_outline():
...
+    [(4, 0, "square"), (8, 0, "octagon"), (4, 45, "square")],
+)
+def test_draw_regular_polygon(n_polygon_sides, rotation, polygon_name):
```
Comment (hugovk) [inline Tests/test_imagedraw.py:1101]:
```suggestion
def test_draw_regular_polygon(n_sides, rotation, polygon_name):
```

File: Tests/test_imagedraw.py [hunk 7]
```diff
@@ -1092,3 +1092,110 @@ def test_same_color_outline():
...
+    im = Image.new("RGBA", size=(W, H), color=(255, 0, 0, 0))
+    filename = (
+        f"Tests/images/imagedraw_regular_polygon__{polygon_name}"
+        f"_with_{rotation}_degree_rotation.png"
+    )
+    draw = ImageDraw.Draw(im)
+    draw.regular_polygon(
+        [(0, 0), (W, H)], n_polygon_sides, rotation=rotation, fill="red"
```
Comment (hugovk) [inline Tests/test_imagedraw.py:1109]:
```suggestion
        [(0, 0), (W, H)], n_sides, rotation=rotation, fill="red"
```

File: Tests/test_imagedraw.py [hunk 8]
```diff
@@ -1092,3 +1092,110 @@ def test_same_color_outline():
...
+)
+def test_compute_regular_polygon_vertices(n_polygon_sides, expected_vertices):
+    vertices = ImageDraw._compute_regular_polygon_vertices(
+        n_sides=n_polygon_sides, bbox=[(0, 0), (100, 100)], rotation=0
+    )
+    assert vertices == expected_vertices
+
+
+@pytest.mark.parametrize(
+    "n_polygon_sides, bounding_box, rotation, expected_error, error_message",
```
Comment (hugovk) [inline Tests/test_imagedraw.py:1150]:
```suggestion
    "n_sides, bbox, rotation, expected_error, error_message",
```

File: Tests/test_imagedraw.py [hunk 9]
```diff
@@ -1092,3 +1092,110 @@ def test_same_color_outline():
...
+            3,
+            [(0, 0), (100, 100)],
+            "0",
+            ValueError,
+            "rotation should be an int or float",
+        ),
+    ],
+)
+def test_compute_regular_polygon_vertices_input_error_handling(
+    n_polygon_sides, bounding_box, rotation, expected_error, error_message
```
Comment (hugovk) [inline Tests/test_imagedraw.py:1195]:
```suggestion
    n_sides, bbox, rotation, expected_error, error_message
```

File: Tests/test_imagedraw.py [hunk 10]
```diff
@@ -1092,3 +1092,110 @@ def test_same_color_outline():
...
+):
+    with pytest.raises(expected_error) as e:
+        ImageDraw._compute_regular_polygon_vertices(
+            n_sides=n_polygon_sides, bbox=bounding_box, rotation=rotation
```
Comment (hugovk) [inline Tests/test_imagedraw.py:1199]:
```suggestion
            n_sides, bbox, rotation
```

File: src/PIL/ImageDraw.py [hunk 11]
```diff
@@ -242,6 +242,13 @@ def polygon(self, xy, fill=None, outline=None):
         if ink is not None and ink != fill:
             self.draw.draw_polygon(xy, ink, 0)
 
+    def regular_polygon(self, bbox, n_sides, rotation=0, fill=None, outline=None):
+        """Draw a regular polygon."""
+        xy = _compute_regular_polygon_vertices(
+            n_sides=n_sides, bbox=bbox, rotation=rotation
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:248]:
```suggestion
            n_sides, bbox, rotation
```

File: src/PIL/ImageDraw.py [hunk 12]
```diff
@@ -242,6 +242,13 @@ def polygon(self, xy, fill=None, outline=None):
...
+        )
+        self.polygon(xy, fill=fill, outline=outline)
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:250]:
```suggestion
        self.polygon(xy, fill, outline)
```

File: src/PIL/ImageDraw.py [hunk 13]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+                current_angle -= 360
+        return angles
+
+    # 3. Variable Declarations
+    vertices = []
+    theta = _get_theta(n_sides=n_sides)
+    side_length = _get_side_length(theta=theta, bbox=bbox)
+    centroid = _get_centroid(bbox=bbox)
+    polygon_radius = _get_polygon_radius(side_length=side_length, theta=theta)
+    angles = _get_angles(n_sides=n_sides, rotation=rotation)
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:695]:
```suggestion
    angles = _get_angles(n_sides, rotation)
```

File: src/PIL/ImageDraw.py [hunk 14]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+            angles.append(current_angle)
+            current_angle += degrees
+            if current_angle > 360:
+                current_angle -= 360
+        return angles
+
+    # 3. Variable Declarations
+    vertices = []
+    theta = _get_theta(n_sides=n_sides)
+    side_length = _get_side_length(theta=theta, bbox=bbox)
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:692]:
```suggestion
    side_length = _get_side_length(theta, bbox)
```

File: src/PIL/ImageDraw.py [hunk 15]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+        for _ in range(0, n_sides):
+            angles.append(current_angle)
+            current_angle += degrees
+            if current_angle > 360:
+                current_angle -= 360
+        return angles
+
+    # 3. Variable Declarations
+    vertices = []
+    theta = _get_theta(n_sides=n_sides)
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:691]:
```suggestion
    theta = _get_theta(n_sides)
```

File: src/PIL/ImageDraw.py [hunk 16]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+    side_length = _get_side_length(theta=theta, bbox=bbox)
+    centroid = _get_centroid(bbox=bbox)
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:693]:
```suggestion
    centroid = _get_centroid(bbox)
```

File: src/PIL/ImageDraw.py [hunk 17]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+    polygon_radius = _get_polygon_radius(side_length=side_length, theta=theta)
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:694]:
```suggestion
    polygon_radius = _get_polygon_radius(side_length, theta)
```

File: src/PIL/ImageDraw.py [hunk 18]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+    angles = _get_angles(n_sides=n_sides, rotation=rotation)
+
+    # 4. Compute Vertices
+    for angle in angles:
+        vertices.append(
+            _compute_polygon_vertex(
+                centroid=centroid, angle=angle, polygon_radius=polygon_radius
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:701]:
```suggestion
               centroid, angle, polygon_radius
```

File: src/PIL/ImageDraw.py [hunk 19]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+
+    def _compute_polygon_vertex(*, angle, centroid, polygon_radius):
+        start_point = [polygon_radius, 0]
+        return _apply_rotation(point=start_point, degrees=angle, centroid=centroid)
+
+    def _get_side_length(*, bbox, theta):
+        h = bbox[3] - bbox[1]
+        return h * math.sin(math.radians(theta))
+
+    def _get_angles(*, n_sides, rotation):
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:677]:
```suggestion
    def _get_angles(n_sides, rotation):
```

File: src/PIL/ImageDraw.py [hunk 20]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+            ),
+        )
+
+    def _get_theta(*, n_sides):
+        return 0.5 * (360 / n_sides)
+
+    def _get_polygon_radius(*, side_length, theta):
+        return (0.5 * side_length) / math.sin(math.radians(theta))
+
+    def _compute_polygon_vertex(*, angle, centroid, polygon_radius):
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:669]:
```suggestion
    def _compute_polygon_vertex(angle, centroid, polygon_radius):
```

File: src/PIL/ImageDraw.py [hunk 21]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+                + point[0] * math.sin(math.radians(360 - degrees))
+                + centroid[1],
+                2,
+            ),
+        )
+
+    def _get_theta(*, n_sides):
+        return 0.5 * (360 / n_sides)
+
+    def _get_polygon_radius(*, side_length, theta):
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:666]:
```suggestion
    def _get_polygon_radius(side_length, theta):
```

File: src/PIL/ImageDraw.py [hunk 22]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+            ),
+            round(
+                point[1] * math.cos(math.radians(360 - degrees))
+                + point[0] * math.sin(math.radians(360 - degrees))
+                + centroid[1],
+                2,
+            ),
+        )
+
+    def _get_theta(*, n_sides):
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:663]:
```suggestion
    def _get_theta(n_sides):
```

File: src/PIL/ImageDraw.py [hunk 23]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+
+    if not bbox[2] - bbox[0] == bbox[3] - bbox[1]:
+        raise ValueError("bbox should be a square")
+
+    # 1.3 Check `rotation` has an appropriate value
+    if not isinstance(rotation, (int, float)):
+        raise ValueError("rotation should be an int or float")
+
+    # 2. Define Helper Functions
+    def _get_centroid(*, bbox):
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:644]:
```suggestion
    def _get_centroid(bbox):
```

File: src/PIL/ImageDraw.py [hunk 24]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+        return (bbox[2] + bbox[0]) * 0.5, (bbox[3] + bbox[1]) * 0.5
+
+    def _apply_rotation(*, point, degrees, centroid):
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:647]:
```suggestion
    def _apply_rotation(point, degrees, centroid):
```

File: src/PIL/ImageDraw.py [hunk 25]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
         edge = new_edge
 
 
+def _compute_regular_polygon_vertices(*, n_sides, bbox, rotation):
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:565]:
```suggestion
def _compute_regular_polygon_vertices(n_sides, bbox, rotation):
```

File: src/PIL/ImageDraw.py [hunk 26]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+    """
+    Generate a list of vertices for a 2D regular polygon.
+
+    :param n_sides: Number of sides
+        (e.g. n_sides = 3 for a triangle, 6 for a hexagon, etc..)
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:570]:
```suggestion
        (e.g. n_sides = 3 for a triangle, 6 for a hexagon)
```

**Review by hugovk — COMMENTED** [review]
Some consistency, nits and simplification!

File: docs/reference/ImageDraw.rst [hunk 27]
```diff
@@ -254,6 +254,22 @@ Methods
...
+    with ``n_sides``, and rotation of ``rotation`` degrees
+
+    :param bbox: A bounding box which inscribes the polygon
+        (e.g. bbox = [(50, 50), (150, 150)])
+    :param n_sides: Number of sides
+        (e.g. n_sides=3 for a triangle, 6 for a hexagon, etc..)
+    :param rotation: Apply an arbitrary rotation to the polygon
+        (e.g. rotation=90, applies a 90 degree rotation)
+    :param outline: Color to use for the outline.
+    :param fill: Color to use for the fill.
```
Comment (hugovk) [inline docs/reference/ImageDraw.rst:270]: Put these in the same order as the signature:

```suggestion
    :param fill: Color to use for the fill.
    :param outline: Color to use for the outline.
```

(I see some of the other parameters in this file are out of order or missing, but we can fix those another time :)

File: src/PIL/ImageDraw.py [hunk 28]
```diff
@@ -555,6 +562,148 @@ def floodfill(image, xy, value, border=None, thresh=0):
...
+        return 0.5 * (360 / n_sides)
+
+    def _get_polygon_radius(*, side_length, theta):
+        return (0.5 * side_length) / math.sin(math.radians(theta))
+
+    def _compute_polygon_vertex(*, angle, centroid, polygon_radius):
+        start_point = [polygon_radius, 0]
+        return _apply_rotation(point=start_point, degrees=angle, centroid=centroid)
+
+    def _get_side_length(*, bbox, theta):
```
Comment (hugovk) [inline src/PIL/ImageDraw.py:673]:
```suggestion
    def _get_side_length(bbox, theta):
```

**Comment by nulano** [comment]
Hmm. The current version draws a polygon inscribed into a circle inscribed into the bounding box. While this is fine for e.g. hexagons, it is not great for squares or triangles:

pr4846-triangle pr4846-square pr4846-hexagon

I implemented my stretching idea, but that is instead not great for hexagons, where it would require users to compute the height for a regular hexagon:

pr4846-stretch-triangle pr4846-stretch-square pr4846-stretch-hexagon

Maybe it would be best to pass the center point in `xy` and a radius instead? That would also solve the problem of what to do for a non-square bbox.

---

For reference, I used the following code:

```python
def test(*args, **kwargs):
     im = Image.new("RGB", (100,100), "yellow")
     d = ImageDraw.Draw(im)
     d.regular_polygon(*args, **kwargs)
     im.show()
test((1,1,98,98), 3, 0, "black")
test((1,1,98,98), 4, 0, "black")
test((1,1,98,98), 6, 0, "black")
```

**Comment by comhar** [comment]
@nulano, thanks for your review.
 
> Maybe it would be best to pass the center point in `xy` and a radius instead? That would also solve the problem of what to do for a non-square bbox.

Agreed. A bounding circle is a good compromise + it significantly simplifies the code.

**Comment by comhar** [comment]
@hugovk, thanks the review. I've implemented your suggestions. Small detail, but what are your thoughts on `b_circle` vs `bcircle`?

**Comment by hugovk** [comment]
Thanks for the updates!

> @hugovk, thanks the review. I've implemented your suggestions. Small detail, but what are your thoughts on `b_circle` vs `bcircle`?

Well, `bbox` is a well-used term, and googling I find helpful results on Wikipedia, OpenStreetMap and R Documentation. Not so with `bcircle`, so I think `b_circle` is better. I'm tempted to suggest `bounding_circle` too (or even just `circle`?), but I don't have a strong opinion.

File: docs/reference/ImageDraw.rst [hunk 29]
```diff
@@ -254,6 +254,22 @@ Methods
...
+
+    Draws a regular polygon inscribed in ``b_circle``,
+    with ``n_sides``, and rotation of ``rotation`` degrees.
+
+    :param b_circle: A bounding circle which inscribes the polygon
+        (e.g. b_circle=[50, 50, 25]).
+    :param n_sides: Number of sides
+        (e.g. n_sides=3 for a triangle, 6 for a hexagon).
+    :param rotation: Apply an arbitrary rotation to the polygon
+        (e.g. rotation=90, applies a 90 degree rotation).
```
Comment (hugovk) [inline docs/reference/ImageDraw.rst:268]:
```suggestion
    :param b_circle: A bounding circle which inscribes the polygon
        (e.g. ``b_circle=[50, 50, 25]``).
    :param n_sides: Number of sides
        (e.g. ``n_sides=3`` for a triangle, ``6`` for a hexagon).
    :param rotation: Apply an arbitrary rotation to the polygon
        (e.g. ``rotation=90``, applies a 90 degree rotation).
```

https://pillow--4846.org.readthedocs.build/en/4846/reference/ImageDraw.html#PIL.ImageDraw.ImageDraw.regular_polygon

**Review by nulano — COMMENTED** [review]
I think `bounding_circle` is *slightly* better than `b_circle`, but both are much better than `bcircle`. Either way, it should accept a point specified as a tuple.

**Comment by comhar** [comment]
> Still not keen on `((x, y), r))`, but I'll defer to others on that.
> 
> Otherwise 👍

Thanks @hugovk , I don't have a strong opinion on this. @nulano, what are your thoughts on removing the `((x, y), r))` formatting option?

**Comment by nulano** [comment]
To me a circle is `(center, radius)`, not `(x, y, r)`.

I can't find other examples online that take a circle parameter, but there doesn't seem to be a strong preference for either:
* separate parameters `xy` and `radius`: https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.patches.Circle.html
* separate parameters `r`, `c`, `radius` (deprecated): https://scikit-image.org/docs/dev/api/skimage.draw.html#skimage.draw.circle
  replaced with parameters `center` and `radius`: https://scikit-image.org/docs/dev/api/skimage.draw.html#skimage.draw.disk
* a class with `Point` and `int` parameters: http://anh.cs.luc.edu/handsonPythonTutorial/graphics.html
* a named tuple of `x y r`: https://books.google.sk/books?id=H9emM_LGFDEC&pg=PA234&lpg=PA234&dq=python+circle+tuple&source=bl&ots=ZbcRbPF8lR&sig=ACfU3U013mMTBOCH_OB7j75t6WB1Uhim0Q&hl=en&sa=X&ved=2ahUKEwjz1fOWsLTrAhUJyqQKHRGvDHUQ6AEwCHoECAEQAQ#v=onepage&q=python circle tuple&f=false

So it is not clear to me that one is better than the other.

---

If only `(x, y, r)` is supported and someone needs to draw polygons with centres stored in a list, they would have to either use the somewhat obscure
```python
for point in points:
    draw.regular_polygon((*point, 5), ...)
```
or pollute the locals with
```python
for x, y in points:
   draw.regular_polygon((x, y, 5), ...)
```

Therefore I think it would be nice to support
```python
for point in points:
    draw.regular_polygon((point, 5), ...)
```
since it doesn't seem to add too much complexity to the function as far as I can tell.

**Comment by hugovk** [comment]
Okay, there's been no comments in a while, so if there's no further suggestions by tomorrow will merge!

**Comment by hugovk** [comment]
@comhar Thank you for your patience and for all your updates, and thank you @nulano for the reviews!

Merging now! Please could you also add release notes in https://github.com/python-pillow/Pillow/blob/master/docs/releasenotes/8.0.0.rst? This can be another PR :)

**Comment by comhar** [comment]
No problem @hugovk , here's the [PR](https://github.com/python-pillow/Pillow/pull/4902) for the release notes.

**Pull request merged**